src/shogun/lib/DynamicArray.h

```diff
@@ -13,30 +13,875 @@
+		 *
+		 * @param p_array another array
+		 * @param dim1 dimension 1
+		 * @param dim2 dimension 2
+		 * @param dim3 dimension 3
+		 * @param p_free_array if array must be freed
+		 * @param p_copy_array if array must be copied
+		 */
+		inline void set_array(
+		    bool* p_array, int32_t dim1, int32_t dim2, int32_t dim3,
+		    bool p_free_array, bool p_copy_array)
+		{
+			dim1_size = dim1;
+			dim2_size = dim2;
+			dim3_size = dim3;
+			int32_t p_array_size = dim1 * dim2 * dim3;
+
+			set_array(p_array, p_array_size, p_free_array, p_copy_array);
+		}
+
+		/** set the array pointer and free previously allocated memory
+		 *
+		 * @param p_array new array
+		 * @param p_num_elements last element index + 1
+		 * @param p_array_size number of elements in array
+		 */
+		inline void set_array(
+		    const bool* p_array, int32_t p_num_elements, int32_t p_array_size)
+		{
+			if (m_array != NULL && free_array)
+				SG_FREE(m_array);
+
+			if (use_sg_mallocs)
+				m_array = SG_MALLOC(bool, p_array_size);
+			else
+				m_array = (bool*)malloc(p_array_size * sizeof(bool));
+			sg_memcpy(m_array, p_array, p_array_size * sizeof(bool));
+
+			array_size = p_array_size;
+			num_elements = p_num_elements;
+			free_array = true;
+		}
+
+		/** clear the array (with e.g. zeros)
+		 * @param value value to fill array with
+		 */
+		inline void clear_array(bool value = false)
+		{
+			if (num_elements - 1 >= 0)
+			{
+				for (int32_t i = 0; i < num_elements; i++)
+					m_array[i] = value;
+			}
+		}
+
+		/** resets the array */
+		inline void reset_array(bool value = false)
+		{
+			clear_array(value);
+			num_elements = 0;
+		}
+
+		/** operator overload for array read only access
+		 * use set_element() for write access (will also make the array
+		 * dynamically grow)
+		 *
+		 * DOES NOT DO ANY BOUNDS CHECKING
+		 *
+		 * @param index index
+		 * @return element at index
+		 */
+		inline const bool& operator[](int32_t index) const
+		{
+			return get_element(index);
+		}
+
+		/** operator overload for array read-write access
+		 *
+		 * DOES NOT DO ANY BOUNDS CHECKING
+		 *
+		 * @param index index
+		 * @return element at index
+		 */
+		inline bool& operator[](int32_t index)
+		{
+			return element(index);
+		}
+
+		/** operator overload for array assignment
+		 *
+		 * @param orig original array
+		 * @return new array
+		 */
+		inline CDynamicArray<bool>& operator=(CDynamicArray<bool>& orig)
+		{
+			m_array = orig.m_array;
+			num_elements = orig.num_elements;
+			dim1_size = orig.dim1_size;
+			dim2_size = orig.dim2_size;
+			dim3_size = orig.dim3_size;
+
+			return *this;
+		}
+
+		/** shuffles the array (not thread safe!) */
+		inline void shuffle()
+		{
+			for (index_t i = 0; i <= num_elements - 1; ++i)
+				CMath::swap(
+				    m_array[i], m_array[CMath::random(i, num_elements - 1)]);
+		}
+
+		/** shuffles the array with external random state */
+		inline void shuffle(CRandom* rand)
```

Comment:
mmm, but after look more into it, I found maybe use `shuffle(sg_rand)` in https://github.com/shogun-toolbox/shogun/blob/f7bccee6251d15f5cfee0514c6d22d3bcf5f7a40/src/shogun/evaluation/StratifiedCrossValidationSplitting.cpp#L155 is a better idea. Sorry for remove it before